// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: market/market.proto

package market

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on CatReq with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *CatReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CatReq with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in CatReqMultiError, or nil if none found.
func (m *CatReq) ValidateAll() error {
	return m.validate(true)
}

func (m *CatReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for CatId

	if len(errors) > 0 {
		return CatReqMultiError(errors)
	}

	return nil
}

// CatReqMultiError is an error wrapping multiple validation errors returned by
// CatReq.ValidateAll() if the designated constraints aren't met.
type CatReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CatReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CatReqMultiError) AllErrors() []error { return m }

// CatReqValidationError is the validation error returned by CatReq.Validate if
// the designated constraints aren't met.
type CatReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CatReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CatReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CatReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CatReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CatReqValidationError) ErrorName() string { return "CatReqValidationError" }

// Error satisfies the builtin error interface
func (e CatReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCatReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CatReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CatReqValidationError{}

// Validate checks the field values on CouponReq with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *CouponReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CouponReq with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in CouponReqMultiError, or nil
// if none found.
func (m *CouponReq) ValidateAll() error {
	return m.validate(true)
}

func (m *CouponReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetCouponId() <= 0 {
		err := CouponReqValidationError{
			field:  "CouponId",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return CouponReqMultiError(errors)
	}

	return nil
}

// CouponReqMultiError is an error wrapping multiple validation errors returned
// by CouponReq.ValidateAll() if the designated constraints aren't met.
type CouponReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CouponReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CouponReqMultiError) AllErrors() []error { return m }

// CouponReqValidationError is the validation error returned by
// CouponReq.Validate if the designated constraints aren't met.
type CouponReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CouponReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CouponReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CouponReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CouponReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CouponReqValidationError) ErrorName() string { return "CouponReqValidationError" }

// Error satisfies the builtin error interface
func (e CouponReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCouponReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CouponReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CouponReqValidationError{}

// Validate checks the field values on CouponUsedReq with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *CouponUsedReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CouponUsedReq with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in CouponUsedReqMultiError, or
// nil if none found.
func (m *CouponUsedReq) ValidateAll() error {
	return m.validate(true)
}

func (m *CouponUsedReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for UserId

	// no validation rules for CouponId

	// no validation rules for OrderId

	if len(errors) > 0 {
		return CouponUsedReqMultiError(errors)
	}

	return nil
}

// CouponUsedReqMultiError is an error wrapping multiple validation errors
// returned by CouponUsedReq.ValidateAll() if the designated constraints
// aren't met.
type CouponUsedReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CouponUsedReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CouponUsedReqMultiError) AllErrors() []error { return m }

// CouponUsedReqValidationError is the validation error returned by
// CouponUsedReq.Validate if the designated constraints aren't met.
type CouponUsedReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CouponUsedReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CouponUsedReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CouponUsedReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CouponUsedReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CouponUsedReqValidationError) ErrorName() string { return "CouponUsedReqValidationError" }

// Error satisfies the builtin error interface
func (e CouponUsedReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCouponUsedReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CouponUsedReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CouponUsedReqValidationError{}

// Validate checks the field values on CouponInfoReq with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *CouponInfoReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CouponInfoReq with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in CouponInfoReqMultiError, or
// nil if none found.
func (m *CouponInfoReq) ValidateAll() error {
	return m.validate(true)
}

func (m *CouponInfoReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for UserId

	// no validation rules for CouponId

	if len(errors) > 0 {
		return CouponInfoReqMultiError(errors)
	}

	return nil
}

// CouponInfoReqMultiError is an error wrapping multiple validation errors
// returned by CouponInfoReq.ValidateAll() if the designated constraints
// aren't met.
type CouponInfoReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CouponInfoReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CouponInfoReqMultiError) AllErrors() []error { return m }

// CouponInfoReqValidationError is the validation error returned by
// CouponInfoReq.Validate if the designated constraints aren't met.
type CouponInfoReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CouponInfoReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CouponInfoReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CouponInfoReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CouponInfoReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CouponInfoReqValidationError) ErrorName() string { return "CouponInfoReqValidationError" }

// Error satisfies the builtin error interface
func (e CouponInfoReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCouponInfoReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CouponInfoReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CouponInfoReqValidationError{}

// Validate checks the field values on HomeCatDataReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *HomeCatDataReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on HomeCatDataReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// HomeCatDataReplyMultiError, or nil if none found.
func (m *HomeCatDataReply) ValidateAll() error {
	return m.validate(true)
}

func (m *HomeCatDataReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Code

	// no validation rules for Message

	for idx, item := range m.GetData() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, HomeCatDataReplyValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, HomeCatDataReplyValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return HomeCatDataReplyValidationError{
					field:  fmt.Sprintf("Data[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return HomeCatDataReplyMultiError(errors)
	}

	return nil
}

// HomeCatDataReplyMultiError is an error wrapping multiple validation errors
// returned by HomeCatDataReply.ValidateAll() if the designated constraints
// aren't met.
type HomeCatDataReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m HomeCatDataReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m HomeCatDataReplyMultiError) AllErrors() []error { return m }

// HomeCatDataReplyValidationError is the validation error returned by
// HomeCatDataReply.Validate if the designated constraints aren't met.
type HomeCatDataReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e HomeCatDataReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e HomeCatDataReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e HomeCatDataReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e HomeCatDataReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e HomeCatDataReplyValidationError) ErrorName() string { return "HomeCatDataReplyValidationError" }

// Error satisfies the builtin error interface
func (e HomeCatDataReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sHomeCatDataReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = HomeCatDataReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = HomeCatDataReplyValidationError{}

// Validate checks the field values on HomeDataReply with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *HomeDataReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on HomeDataReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in HomeDataReplyMultiError, or
// nil if none found.
func (m *HomeDataReply) ValidateAll() error {
	return m.validate(true)
}

func (m *HomeDataReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Code

	// no validation rules for Message

	for idx, item := range m.GetData() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, HomeDataReplyValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, HomeDataReplyValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return HomeDataReplyValidationError{
					field:  fmt.Sprintf("Data[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return HomeDataReplyMultiError(errors)
	}

	return nil
}

// HomeDataReplyMultiError is an error wrapping multiple validation errors
// returned by HomeDataReply.ValidateAll() if the designated constraints
// aren't met.
type HomeDataReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m HomeDataReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m HomeDataReplyMultiError) AllErrors() []error { return m }

// HomeDataReplyValidationError is the validation error returned by
// HomeDataReply.Validate if the designated constraints aren't met.
type HomeDataReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e HomeDataReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e HomeDataReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e HomeDataReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e HomeDataReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e HomeDataReplyValidationError) ErrorName() string { return "HomeDataReplyValidationError" }

// Error satisfies the builtin error interface
func (e HomeDataReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sHomeDataReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = HomeDataReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = HomeDataReplyValidationError{}

// Validate checks the field values on AppSettingReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *AppSettingReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on AppSettingReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// AppSettingReplyMultiError, or nil if none found.
func (m *AppSettingReply) ValidateAll() error {
	return m.validate(true)
}

func (m *AppSettingReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Code

	// no validation rules for Message

	for idx, item := range m.GetData() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, AppSettingReplyValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, AppSettingReplyValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return AppSettingReplyValidationError{
					field:  fmt.Sprintf("Data[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return AppSettingReplyMultiError(errors)
	}

	return nil
}

// AppSettingReplyMultiError is an error wrapping multiple validation errors
// returned by AppSettingReply.ValidateAll() if the designated constraints
// aren't met.
type AppSettingReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m AppSettingReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m AppSettingReplyMultiError) AllErrors() []error { return m }

// AppSettingReplyValidationError is the validation error returned by
// AppSettingReply.Validate if the designated constraints aren't met.
type AppSettingReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e AppSettingReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e AppSettingReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e AppSettingReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e AppSettingReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e AppSettingReplyValidationError) ErrorName() string { return "AppSettingReplyValidationError" }

// Error satisfies the builtin error interface
func (e AppSettingReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sAppSettingReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = AppSettingReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = AppSettingReplyValidationError{}

// Validate checks the field values on NoticeReply with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *NoticeReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on NoticeReply with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in NoticeReplyMultiError, or
// nil if none found.
func (m *NoticeReply) ValidateAll() error {
	return m.validate(true)
}

func (m *NoticeReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Code

	// no validation rules for Message

	for idx, item := range m.GetData() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, NoticeReplyValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, NoticeReplyValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return NoticeReplyValidationError{
					field:  fmt.Sprintf("Data[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return NoticeReplyMultiError(errors)
	}

	return nil
}

// NoticeReplyMultiError is an error wrapping multiple validation errors
// returned by NoticeReply.ValidateAll() if the designated constraints aren't met.
type NoticeReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m NoticeReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m NoticeReplyMultiError) AllErrors() []error { return m }

// NoticeReplyValidationError is the validation error returned by
// NoticeReply.Validate if the designated constraints aren't met.
type NoticeReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NoticeReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NoticeReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NoticeReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NoticeReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NoticeReplyValidationError) ErrorName() string { return "NoticeReplyValidationError" }

// Error satisfies the builtin error interface
func (e NoticeReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNoticeReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NoticeReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NoticeReplyValidationError{}

// Validate checks the field values on SearchReply with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *SearchReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SearchReply with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in SearchReplyMultiError, or
// nil if none found.
func (m *SearchReply) ValidateAll() error {
	return m.validate(true)
}

func (m *SearchReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Code

	// no validation rules for Message

	for idx, item := range m.GetData() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, SearchReplyValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, SearchReplyValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return SearchReplyValidationError{
					field:  fmt.Sprintf("Data[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return SearchReplyMultiError(errors)
	}

	return nil
}

// SearchReplyMultiError is an error wrapping multiple validation errors
// returned by SearchReply.ValidateAll() if the designated constraints aren't met.
type SearchReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SearchReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SearchReplyMultiError) AllErrors() []error { return m }

// SearchReplyValidationError is the validation error returned by
// SearchReply.Validate if the designated constraints aren't met.
type SearchReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SearchReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SearchReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SearchReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SearchReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SearchReplyValidationError) ErrorName() string { return "SearchReplyValidationError" }

// Error satisfies the builtin error interface
func (e SearchReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSearchReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SearchReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SearchReplyValidationError{}

// Validate checks the field values on PayReply with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *PayReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on PayReply with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in PayReplyMultiError, or nil
// if none found.
func (m *PayReply) ValidateAll() error {
	return m.validate(true)
}

func (m *PayReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Code

	// no validation rules for Message

	for idx, item := range m.GetData() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, PayReplyValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, PayReplyValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return PayReplyValidationError{
					field:  fmt.Sprintf("Data[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return PayReplyMultiError(errors)
	}

	return nil
}

// PayReplyMultiError is an error wrapping multiple validation errors returned
// by PayReply.ValidateAll() if the designated constraints aren't met.
type PayReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m PayReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m PayReplyMultiError) AllErrors() []error { return m }

// PayReplyValidationError is the validation error returned by
// PayReply.Validate if the designated constraints aren't met.
type PayReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e PayReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e PayReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e PayReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e PayReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e PayReplyValidationError) ErrorName() string { return "PayReplyValidationError" }

// Error satisfies the builtin error interface
func (e PayReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sPayReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = PayReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = PayReplyValidationError{}

// Validate checks the field values on CouponListReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *CouponListReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CouponListReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CouponListReplyMultiError, or nil if none found.
func (m *CouponListReply) ValidateAll() error {
	return m.validate(true)
}

func (m *CouponListReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Code

	// no validation rules for Message

	for idx, item := range m.GetData() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, CouponListReplyValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, CouponListReplyValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return CouponListReplyValidationError{
					field:  fmt.Sprintf("Data[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return CouponListReplyMultiError(errors)
	}

	return nil
}

// CouponListReplyMultiError is an error wrapping multiple validation errors
// returned by CouponListReply.ValidateAll() if the designated constraints
// aren't met.
type CouponListReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CouponListReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CouponListReplyMultiError) AllErrors() []error { return m }

// CouponListReplyValidationError is the validation error returned by
// CouponListReply.Validate if the designated constraints aren't met.
type CouponListReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CouponListReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CouponListReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CouponListReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CouponListReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CouponListReplyValidationError) ErrorName() string { return "CouponListReplyValidationError" }

// Error satisfies the builtin error interface
func (e CouponListReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCouponListReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CouponListReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CouponListReplyValidationError{}
